Mathlib/Topology/CompletelyRegular.lean

```diff
@@ -0,0 +1,104 @@
+
+## Implementation notes
+
+The present definition `CompletelyRegularSpace` is a slight modification of the one given in
+[russell1974]. There it's assumed that any point `x ∈ Kᶜ` is separated from the closed set `K` by a
+continuous *real* valued function `f` (as opposed to `f` being unit-interval-valued). This can be
+converted to the present definition by replacing a real-valued `f` by `h ∘ g ∘ f`, with
+`g : x ↦ max(x, 0)` and `h : x ↦ min(x, 1)`.
+
+## References
+
+* [Russell C. Walker, *The Stone-Čech Compactification*][russell1974]
+-/
+
+universe u
+
+noncomputable section
+
+open Filter Set Topology unitInterval Set.Icc
+
+variable {X : Type u} [TopologicalSpace X] [T1Space X]
+
+/-- A T1 space X is completely regular if points can be separated from closed sets via
+  continuous functions to the unit interval. -/
+@[mk_iff completelyRegularSpace_iff]
+class CompletelyRegularSpace (X : Type u) [TopologicalSpace X] [T1Space X] : Prop where
+  completely_regular : ∀ (x : X), ∀ (K : Set X) (_: IsClosed K), x ∉ K →
+    ∃ f : X → I, Continuous f ∧ f x = 0 ∧ EqOn f 1 K
+
+instance CompletelyRegularSpace.t3space [CompletelyRegularSpace X] : T3Space X := by
+  have : T0Space X := by apply T1Space.t0Space
+  have : RegularSpace X := by
+    have ne : (0 : I) ≠ 1 := by
+      apply Icc.coe_ne_one.1
+      rw [coe_zero]
+      apply zero_ne_one
+    rw [regularSpace_iff]
+    intro s a hs ha
+    obtain ⟨f, cf, hf, hhf⟩ := CompletelyRegularSpace.completely_regular a s hs ha
+    rw [Filter.disjoint_iff]
+    let ⟨u, v, hu, hv, hhu, hhv, huv⟩ := T2Space.t2 (0 : I) 1 ne
+    have mem_u : f ⁻¹' u ∈ (𝓝 a) := by
+      have L : a ∈ ↑f ⁻¹' u := by rwa [mem_preimage, hf]
+      rw [mem_nhds_iff]
+      exact ⟨f ⁻¹' u, Subset.refl (f ⁻¹' u), IsOpen.preimage cf hu, L⟩
+    have mem_v : f ⁻¹' v ∈ (𝓝ˢ s) := by
+      have vo := IsOpen.preimage cf hv
+      rw [←interior_eq_iff_isOpen] at vo
+      rw [←subset_interior_iff_mem_nhdsSet, vo]
+      intro x hx
+      have f_x : f x = 1 := by exact hhf hx
+      rwa [mem_preimage, f_x]
+    have duv : Disjoint (f ⁻¹' v) (f ⁻¹' u) := by
+      rw [Set.disjoint_iff, inter_comm, subset_empty_iff] at huv
+      rw [Set.disjoint_iff, ←preimage_inter, huv, preimage_empty]
+    exact ⟨f ⁻¹' v, mem_v, f ⁻¹' u, mem_u, duv⟩
+  exact {}
+
+instance NormalSpace.completelyRegularSpace [NormalSpace X] : CompletelyRegularSpace X := by
+  rw [completelyRegularSpace_iff]
+  intro x K hK hx
+  have cx : IsClosed {x} := by apply T1Space.t1
+  have d : Disjoint {x} K := by rwa [Set.disjoint_iff, subset_empty_iff, singleton_inter_eq_empty]
+  let ⟨⟨f, cf⟩, hfx, hfK, hficc⟩ := exists_continuous_zero_one_of_closed cx hK d
+  let g : X → I := fun x => ⟨f x, hficc x⟩
+  have cg : Continuous g := cf.subtype_mk hficc
+  have hgx : g x = ⟨0, zero_mem⟩ := by
+    rw [mk_zero, ←coe_eq_zero]
+    apply hfx
+    rw [mem_singleton_iff]
+  have hgK : EqOn g 1 K := by
+    intro k
+    rw [Pi.one_apply, ←coe_eq_one]
+    apply hfK
```

Comment:
Something like `fun k hk => Subtype.ext (hfK k hk)`
BTW you can Shift-select multiple lines when adding comments.